stratum/hal/lib/phal/onlp/onlp_wrapper.cc

```diff
@@ -56,28 +57,122 @@ OnlpWrapper* OnlpWrapper::CreateSingleton() {
   absl::WriterMutexLock l(&init_lock_);
   if (!singleton_) {
     singleton_ = new OnlpWrapper();
+    ::util::Status status = singleton_->Init();
+    if (!status.ok()) {
+      LOG(ERROR) << "OnlpWrapper::Init() failed: " << status;
+      delete singleton_;
+      singleton_ = nullptr;
+    }
   }
 
   return singleton_;
 }
 
-OnlpWrapper::OnlpWrapper() {
-  LOG(INFO) << "Initializing ONLP.";
-  if (ONLP_FAILURE(onlp_sw_init(nullptr))) {
-    LOG(FATAL) << "Failed to initialize ONLP.";
-  }
-}
+OnlpWrapper::OnlpWrapper() {}
 
 OnlpWrapper::~OnlpWrapper() {
   LOG(INFO) << "Deinitializing ONLP.";
-  if (ONLP_FAILURE(onlp_sw_denit())) {
+  if (ONLP_FAILURE(onlp_functions_.onlp_sw_denit())) {
     LOG(ERROR) << "Failed to deinitialize ONLP.";
   }
+  // TODO(max): log error
+  dlclose(onlp_lib_handle_);
+}
+
+namespace {
+template <typename T>
+::util::StatusOr<T> load_symbol(void* handle, const char* name) {
+  dlerror();  // Clear last error.
+  auto* symbol = reinterpret_cast<T>(dlsym(handle, name));
+  char* dl_err = dlerror();
+  CHECK_RETURN_IF_FALSE(dl_err == nullptr)
+      << "Failed to load symbol " << name << ": " << dl_err;
+
+  return symbol;
+}
+}  // namespace
+
+::util::Status OnlpWrapper::Init() {
+  LOG(INFO) << "Initializing ONLP.";
+
+  onlp_lib_handle_ = dlopen("libonlp.so", RTLD_NOW);
+  CHECK_RETURN_IF_FALSE(onlp_lib_handle_ != nullptr)
+      << "Failed to open libonlp.so: " << dlerror();
+
+  // TODO(max): simplify with macro or something.
```

Comment:
With some string tokenization this could be `LOAD_SYMBOL("foo")`.

Reply:
local macro makes sense